src/rust/ide/src/controller/visualization.rs

```diff
@@ -0,0 +1,254 @@
+//! Visualization controller.
+//!
+//! Ths Visualization Controller is Responsible identifying all the available visualization natively
+//! embedded in IDE and available within the project's `visualization` folder. The
+//! `Visualization Controller` lives as long as the `Project Controller`.
+
+use crate::prelude::*;
+
+use std::rc::Rc;
+use enso_protocol::language_server;
+use graph_editor::component::visualization::class;
+use graph_editor::component::visualization::JsSourceClass;
+
+
+
+// =============
+// === Error ===
+// =============
+
+/// Enumeration of errors used in `Visualization Controller`.
+#[derive(Debug,Fail)]
+#[allow(missing_docs)]
+pub enum VisualizationError {
+    #[fail(display = "Visualization \"{}\" not found.", identifier)]
+    NotFound {
+        identifier : VisualizationIdentifier
+    },
+    #[fail(display = "JavaScript visualization \"{}\" failed to compile.", identifier)]
+    CompileError {
+        identifier : VisualizationIdentifier
+    }
+}
+
+
+
+// ===============================
+// === VisualizationIdentifier ===
+// ===============================
+
+/// This enum is used to identify a visualization both in the project folder or natively embedded in
+/// IDE.
+#[derive(Clone,Debug,Display,Eq,PartialEq)]
+#[allow(missing_docs)]
+pub enum VisualizationIdentifier {
+    Embedded(String),
+    File(language_server::Path)
+}
+
+
+
+// ==============================
+// === EmbeddedVisualizations ===
+// ==============================
+
+#[allow(missing_docs)]
+pub type EmbeddedVisualizationName   = String;
+
+/// Embedded visualizations mapped from name to source code.
+#[derive(Shrinkwrap,Debug,Clone,Default)]
+#[shrinkwrap(mutable)]
+pub struct EmbeddedVisualizations {
+    #[allow(missing_docs)]
+    pub map : HashMap<EmbeddedVisualizationName,Rc<class::Handle>>
+}
+
+
+
+
+// ==============
+// === Handle ===
+// ==============
+
+const VISUALISATION_FOLDER : &str = "visualization";
+
+#[derive(Debug)]
+struct VisualizationControllerData {
+    language_server_rpc     : Rc<language_server::Connection>,
+    embedded_visualizations : EmbeddedVisualizations
+}
+
+/// Visualization Controller's state.
+#[derive(Debug,Clone,CloneRef)]
+pub struct Handle {
+    data : Rc<RefCell<VisualizationControllerData>>
+}
+
+impl Handle {
+    /// Creates a new visualization controller.
+    pub fn new
+    ( language_server_rpc     : Rc<language_server::Connection>
+    , embedded_visualizations : EmbeddedVisualizations) -> Self {
+        let data = VisualizationControllerData{language_server_rpc,embedded_visualizations};
+        let data = Rc::new(RefCell::new(data));
+        Self {data}
+    }
+
+    async fn list_file_visualizations(&self) -> FallibleResult<Vec<VisualizationIdentifier>> {
```

Comment:
I'd consider naming this more like "list_project_specific_visualizations".
I expect that embedded visualizations can (or in future will can) be recognized as files as well.